File: swift/Sources/benchmarks/GameOfLife.swift
import Foundation

final class GameOfLife: BenchmarkProtocol {
    private final class Cell {
        var alive: Bool = false
        var nextState: Bool = false
        var neighbors: [Cell] = []

        func addNeighbor(_ cell: Cell) {
            neighbors.append(cell)
        }

        func computeNextState() {
            var aliveNeighbors = 0
            for neighbor in neighbors {
                if neighbor.alive { aliveNeighbors += 1 }
            }

            if alive {
                nextState = aliveNeighbors == 2 || aliveNeighbors == 3
            } else {
                nextState = aliveNeighbors == 3
            }
        }

        func update() {
            alive = nextState
        }
    }

    private final class Grid {
        let width: Int
        let height: Int
        private var cells: [[Cell]]

        init(width: Int, height: Int) {
            self.width = width
            self.height = height

            cells = (0..<height).map { y in
                (0..<width).map { x in
                    Cell()
                }
            }

            linkNeighbors()
        }

        private func linkNeighbors() {
            for y in 0..<height {
                for x in 0..<width {
                    let cell = cells[y][x]

                    for dy in -1...1 {
                        for dx in -1...1 {
                            if dx == 0 && dy == 0 { continue }

                            let ny = (y + dy + height) % height
                            let nx = (x + dx + width) % width

                            cell.addNeighbor(cells[ny][nx])
                        }
                    }
                }
            }
        }

        func nextGeneration() {

            for row in cells {
                for cell in row {
                    cell.computeNextState()
                }
            }

            for row in cells {
                for cell in row {
                    cell.update()
                }
            }
        }

        func countAlive() -> Int {
            return cells.reduce(0) { total, row in
                total + row.filter { $0.alive }.count
            }
        }

        func computeHash() -> UInt32 {
            let FNV_OFFSET_BASIS: UInt32 = 2166136261
            let FNV_PRIME: UInt32 = 16777619

            var hasher = FNV_OFFSET_BASIS

            for row in cells {
                for cell in row {
                    let alive = cell.alive ? UInt32(1) : UInt32(0)
                    hasher ^= alive
                    hasher = hasher &* FNV_PRIME
                }
            }

            return hasher
        }

        subscript(x: Int, y: Int) -> Cell {
            return cells[y][x]
        }
    }

    private var width: Int = 0
    private var height: Int = 0
    private var grid: Grid

    init() {
        self.width = 0
        self.height = 0
        self.grid = Grid(width: 0, height: 0)
    }

    func prepare() {
        let configW = Int(configValue("w") ?? 256)
        let configH = Int(configValue("h") ?? 256)

        self.width = configW
        self.height = configH
        self.grid = Grid(width: width, height: height)

        for y in 0..<height {
            for x in 0..<width {
                if Helper.nextFloat() < 0.1 {
                    grid[x, y].alive = true
                }
            }
        }
    }

    func run(iterationId: Int) {
        grid.nextGeneration()
    }

    var checksum: UInt32 {
        let alive = grid.countAlive()
        return grid.computeHash() + UInt32(alive)
    }

    var name: String {
        return "GameOfLife"
    }
}

File: fsharp/GameOfLife.fs
namespace Benchmarks

open System

type Cell() =
    member val Alive = false with get, set
    member val NextState = false with get, set
    member val Neighbors = Array.zeroCreate<Cell> 8 with get
    member val NeighborCount = 0 with get, set

    member this.AddNeighbor(cell: Cell) =
        this.Neighbors.[this.NeighborCount] <- cell
        this.NeighborCount <- this.NeighborCount + 1

    member this.ComputeNextState() =
        let mutable aliveNeighbors = 0
        for i in 0 .. this.NeighborCount - 1 do
            if this.Neighbors.[i].Alive then
                aliveNeighbors <- aliveNeighbors + 1

        this.NextState <-
            if this.Alive then
                aliveNeighbors = 2 || aliveNeighbors = 3
            else
                aliveNeighbors = 3

    member this.Update() =
        this.Alive <- this.NextState

type Grid(width: int, height: int) =
    let cells = Array2D.init height width (fun _ _ -> Cell())

    do

        for y in 0 .. height - 1 do
            for x in 0 .. width - 1 do
                let cell = cells.[y, x]
                for dy in -1 .. 1 do
                    for dx in -1 .. 1 do
                        if not (dx = 0 && dy = 0) then
                            let ny = (y + dy + height) % height
                            let nx = (x + dx + width) % width
                            cell.AddNeighbor(cells.[ny, nx])

    member _.Width = width
    member _.Height = height

    member this.NextGeneration() =

        for y in 0 .. height - 1 do
            for x in 0 .. width - 1 do
                cells.[y, x].ComputeNextState()

        for y in 0 .. height - 1 do
            for x in 0 .. width - 1 do
                cells.[y, x].Update()

        this

    member _.CountAlive() =
        let mutable count = 0
        for y in 0 .. height - 1 do
            for x in 0 .. width - 1 do
                if cells.[y, x].Alive then
                    count <- count + 1
        count

    member _.ComputeHash() =
        let mutable hash = 2166136261u
        for y in 0 .. height - 1 do
            for x in 0 .. width - 1 do
                let alive = if cells.[y, x].Alive then 1u else 0u
                hash <- (hash ^^^ alive) * 16777619u
        hash

    member _.GetCells() = cells

type GameOfLife() =
    inherit Benchmark()

    let mutable grid : Grid option = None

    override this.Checksum =
        match grid with
        | Some g -> g.ComputeHash() + uint32 (g.CountAlive())
        | None -> 0u

    override this.Prepare() =
        let width = Helper.Config_i64("GameOfLife", "w") |> int
        let height = Helper.Config_i64("GameOfLife", "h") |> int

        let g = Grid(width, height)

        for y in 0 .. height - 1 do
            for x in 0 .. width - 1 do
                if Helper.NextFloat(1.0) < 0.1 then
                    g.GetCells().[y, x].Alive <- true

        grid <- Some g

    override this.Run(IterationId: int64) =
        match grid with
        | Some g -> g.NextGeneration() |> ignore
        | None -> ()